stylus-sdk/src/abi/ints.rs

```diff
@@ -0,0 +1,238 @@
+// Copyright 2023-2024, Offchain Labs, Inc.
+// For licensing, see https://github.com/OffchainLabs/stylus-sdk-rs/blob/stylus/licenses/COPYRIGHT.md
+
+//! Support for generic integer types, found in [alloy_primitives].
+
+use alloy_primitives::{ruint::UintTryFrom, Signed, Uint};
+use alloy_sol_types::{
+    abi::token::WordToken,
+    private::SolTypeValue,
+    sol_data::{self, IntBitCount, SupportedInt},
+    SolType,
+};
+
+use super::{AbiType, ConstString};
+
+/// Const lookup table for int type ABI names
+const INT_ABI_LOOKUP: [&'static str; 32] = [
+    "int8", "int16", "int24", "int32", "int40", "int48", "int56", "int64", "int72", "int80",
+    "int88", "int96", "int104", "int112", "int120", "int128", "int136", "int144", "int152",
+    "int160", "int168", "int176", "int184", "int192", "int200", "int208", "int216", "int224",
+    "int232", "int240", "int248", "int256",
+];
+
+/// Represents [`intX`] in Solidity
+///
+/// A custom type is used here in lieu of [alloy_sol_types::sol_data::Int] in order to keep our
+/// bijectivity constraint.
+///
+/// [`intX`]: https://docs.soliditylang.org/en/latest/types.html#integers
+pub struct SolInt<const BITS: usize, const LIMBS: usize>;
+
+impl<const BITS: usize, const LIMBS: usize> AbiType for Signed<BITS, LIMBS>
+where
```

Comment:
This is the most hardcore generic constraint I've seen 😅, great job on getting it to work